### PR TITLE
Adding OCIS setup to work with LLDAP

### PR DESCRIPTION
**Pull request by adamshand**
Tested and works.

### CODE REVIEW

File: example_configs/ocis [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      OCIS_LDAP_BIND_DN: "uid=admin,ou=people,${LDAP_BASE_DN}"
+      OCIS_LDAP_BIND_PASSWORD: ${LDAP_BIND_PASSWORD}
+      OCIS_ADMIN_USER_ID: ${LLDAP_ADMIN_UUID}
+
+      OCIS_LDAP_USER_ENABLED_ATTRIBUTE: uid
+      GRAPH_LDAP_SERVER_WRITE_ENABLED: "false" # Does your LLDAP bind user have write access?
+      GRAPH_LDAP_REFINT_ENABLED: "false"
+      # Disable the built in LDAP server
+      OCIS_EXCLUDE_RUN_SERVICES: idm
+      # both text and binary cause errors in LLDAP, seems harmless though (?)
```
Comment (nitnelave) [inline example_configs/ocis:69]: What's the error you see? I'm curious

File: example_configs/ocis [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      # can filter which groups are imported, eg: `(&(objectclass=groupOfUniqueNames)(uid=ocis_*))`
+      OCIS_LDAP_GROUP_FILTER: "(objectclass=groupOfUniqueNames)"
+
+      OCIS_LDAP_USER_BASE_DN: "ou=people,${LDAP_BASE_DN}"
+      OCIS_LDAP_USER_OBJECTCLASS: "inetOrgPerson"
+      # Allows all users
+      #OCIS_LDAP_USER_FILTER: "(objectclass=inetOrgPerson)"
+      # Allows users who are in the LLDAP group `ocis_users`
+      OCIS_LDAP_USER_FILTER: "(&(objectclass=person)(memberOf=cn=ocis_users,ou=groups,${LDAP_BASE_DN}))"
+      # NOT WORKING: Used instead of restricting users with OCIS_LDAP_USER_FILTER
```
Comment (nitnelave) [inline example_configs/ocis:90]: Curious that it's not working, it looks fairly straightforward. Do you have any details?

File: example_configs/ocis (hunk 1 above)
Comment (adamshand) [reply example_configs/ocis:69]:
```
lldap  | 2024-02-19T05:47:32.478666428Z 2024-02-19T05:47:32.478151155+00:00  WARN        ┕━ 🚧 [warn]: Ignoring unrecognized group attribute: text\n\
lldap  | 2024-02-19T05:47:32.478673132Z                       To disable this warning, add it to "ignored_user_attributes" in the config.
```

As far as I can tell from the docs, there's only two possible options `text` and `binary`.  Both generate the same error, but `binary` doesn't work for auth.

File: example_configs/ocis (hunk 2 above)
Comment (adamshand) [reply example_configs/ocis:90]: Can't remember, let me try again and remember ...

If I change ...

```
OCIS_LDAP_USER_FILTER: "(objectclass=inetOrgPerson)"
#OCIS_LDAP_USER_FILTER: "(&(objectclass=person)(memberOf=cn=ocis_users,ou=groups,${LDAP_BASE_DN}))"
OCIS_LDAP_DISABLE_USER_MECHANISM: "group"
OCIS_LDAP_DISABLED_USERS_GROUP_DN: "uid=ocis_disabled,ou=groups,${LDAP_BASE_DN}"
```

I get this over and over in the OCIS logs.  I can log into OCIS, but the web app hangs on a splash screen.  The `ocis_disabled` group already exists in LLDAP.

```
ocis  | 2024-02-19T07:00:57.191571167Z {"level":"error","service":"graph","error":"could not initialize graph service","transport":"http","time":"2024-02-19T07:00:57Z","message":"Failed to initialize server"}
ocis  | 2024-02-19T07:00:57.278463335Z {"level":"error","service":"graph","error":"LDAP Result Code 34 \"Invalid DN Syntax\": Unexpected DN format. Got \"uid=ocis_disabled,ou=groups,dc=example,dc=nz\", expected: \"uid=id,ou=people,dc=example,dc=nz\"","local_user_disable":"uid=ocis_disabled,ou=groups,dc=example,dc=nz","time":"2024-02-19T07:00:57Z","message":"error adding group for disabling users"}
```

Not sure what's up with `Unexpected DN format. Got \"uid=ocis_disabled,ou=groups,dc=example,dc=nz\", expected: \"uid=id,ou=people,dc=example,dc=nz\"`

There aren't any useful logs in LLDAP. 

Reading that error again I see it's trying to write to LLDAP.  Changing ...

```
+      GRAPH_LDAP_SERVER_WRITE_ENABLED: "true" # assuming the external ldap is writable
```

Doesn't seem to make a difference.

File: example_configs/ocis (hunk 1 above)
Comment (nitnelave) [reply example_configs/ocis:69]: Can you enable verbose mode? I'd like to see the LDAP query that results into this. It seems like it's just listing "text" as an extra attribute, I wonder how they expect it to work.

File: example_configs/ocis (hunk 2 above)
Comment (nitnelave) [reply example_configs/ocis:90]: LLDAP would have more useful logs with verbose mode enabled.

File: example_configs/ocis (hunk 1 above)
Comment (adamshand) [reply example_configs/ocis:69]:
```
lldap  | 2024-02-19T18:56:32.279688737Z 2024-02-19T18:56:32.189192027+00:00  INFO     ┝━ LDAP request [ 1.43ms | 0.10% / 0.67% ]
lldap  | 2024-02-19T18:56:32.279694940Z 2024-02-19T18:56:32.189211273+00:00  DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 2, op: SearchRequest(LdapSearchRequest { base: "ou=people,dc=example,dc=nz", scope: Subtree, aliases: Never, sizelimit: 1, timelimit: 60, typesonly: false, filter: And([And([Equality("objectclass", "inetOrgPerson"), And([Equality("objectclass", "person"), Equality("memberOf", "cn=ocis_users,ou=groups,dc=example,dc=nz")]), Not(Equality("uid", "FALSE"))]), Equality("uid", "adam")]), attrs: ["dn", "uid", "mail", "displayName", "sn", "givenName", "entryuuid", "text"] }), ctrl: [] }
lldap  | 2024-02-19T18:56:32.279729418Z 2024-02-19T18:56:32.189213744+00:00  DEBUG    │  ┝━ do_search [ 1.22ms | 0.13% / 0.57% ]
lldap  | 2024-02-19T18:56:32.279737489Z 2024-02-19T18:56:32.189223596+00:00  DEBUG    │  │  ┝━ 🐛 [debug]:  | request.base: "ou=people,dc=example,dc=nz" | scope: Users
lldap  | 2024-02-19T18:56:32.279744623Z 2024-02-19T18:56:32.189226524+00:00  DEBUG    │  │  ┝━ get_user_list [ 935µs | 0.02% / 0.44% ]
lldap  | 2024-02-19T18:56:32.279750738Z 2024-02-19T18:56:32.189257385+00:00  DEBUG    │  │  │  ┝━ 🐛 [debug]:  | filters: And([And([And([]), And([And([]), MemberOf("ocis_users")]), Not(UserId(UserId("false")))]), UserId(UserId("adam"))])
lldap  | 2024-02-19T18:56:32.279757960Z 2024-02-19T18:56:32.189265069+00:00  DEBUG    │  │  │  ┕━ list_users [ 885µs | 0.42% ] filters: Some(And([And([And([]), And([And([]), MemberOf("ocis_users")]), Not(UserId(UserId("false")))]), UserId(UserId("adam"))])) | _get_groups: false
lldap  | 2024-02-19T18:56:32.279768669Z 2024-02-19T18:56:32.190937607+00:00  DEBUG    │  │  │     ┕━ 🐛 [debug]:  | return: [UserAndGroups { user: User { user_id: UserId("adam"), email: "adam@example.nz", display_name: Some("Adam Shand"), creation_date: 2023-05-19T09:37:38.282695250, uuid: Uuid("c5c8724a-bfaa-36d0-ae77-6cc936bf5808"), attributes: [AttributeValue { name: "avatar", value: Serialized("hash: 0xEDB02F50A5FB7F1D") }, AttributeValue { name: "first_name", value: Serialized("Adam") }, AttributeValue { name: "last_name", value: Serialized("Shand") }] }, groups: Some([GroupDetails { group_id: GroupId(5), display_name: "calibre", creation_date: 2023-05-20T05:15:14.956410978, uuid: Uuid("59463dc6-5700-3d0d-96bc-432a5aa43f5f"), attributes: [] }, GroupDetails { group_id: GroupId(4), display_name: "jellyfin-admin", creation_date: 2023-05-20T03:35:06.456309899, uuid: Uuid("e5c0de8a-25a7-37f1-be06-eb82b97e7d6d"), attributes: [] }, GroupDetails { group_id: GroupId(1), display_name: "lldap_admin", creation_date: 2023-05-19T09:35:39.078795764, uuid: Uuid("100fcf69-388b-3725-889f-8c496f9cec40"), attributes: [] }, GroupDetails { group_id: GroupId(6), display_name: "ocis_users", creation_date: 2024-02-19T01:08:57.207719922, uuid: Uuid("4afbf888-08c1-336a-8c79-45a7c4570060"), attributes: [] }]) }]
lldap  | 2024-02-19T18:56:32.279785740Z 2024-02-19T18:56:32.191568379+00:00  DEBUG    │  │  ┝━ expand_attribute_wildcards [ 13.3µs | 0.01% ] ldap_attributes: ["dn", "uid", "mail", "displayName", "sn", "givenName", "entryuuid", "text"]
lldap  | 2024-02-19T18:56:32.279792693Z 2024-02-19T18:56:32.191581211+00:00  DEBUG    │  │  │  ┕━ 🐛 [debug]:  | resolved_attributes: ["dn", "uid", "mail", "displayName", "sn", "givenName", "entryuuid", "text"]
lldap  | 2024-02-19T18:56:32.279798975Z 2024-02-19T18:56:32.191595559+00:00  WARN     │  │  ┕━ 🚧 [warn]: Ignoring unrecognized group attribute: text\n\
lldap  | 2024-02-19T18:56:32.279805325Z                       To disable this warning, add it to "ignored_user_attributes" in the config.
lldap  | 2024-02-19T18:56:32.279688737Z 2024-02-19T18:56:32.189192027+00:00  INFO     ┝━ LDAP request [ 1.43ms | 0.10% / 0.67% ]
lldap  | 2024-02-19T18:56:32.279694940Z 2024-02-19T18:56:32.189211273+00:00  DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 2, op: SearchRequest(LdapSearchRequest { base: "ou=people,dc=example,dc=nz", scope: Subtree, aliases: Never, sizelimit: 1, timelimit: 60, typesonly: false, filter: And([And([Equality("objectclass", "inetOrgPerson"), And([Equality("objectclass", "person"), Equality("memberOf", "cn=ocis_users,ou=groups,dc=example,dc=nz")]), Not(Equality("uid", "FALSE"))]), Equality("uid", "adam")]), attrs: ["dn", "uid", "mail", "displayName", "sn", "givenName", "entryuuid", "text"] }), ctrl: [] }
lldap  | 2024-02-19T18:56:32.279729418Z 2024-02-19T18:56:32.189213744+00:00  DEBUG    │  ┝━ do_search [ 1.22ms | 0.13% / 0.57% ]
lldap  | 2024-02-19T18:56:32.279737489Z 2024-02-19T18:56:32.189223596+00:00  DEBUG    │  │  ┝━ 🐛 [debug]:  | request.base: "ou=people,dc=example,dc=nz" | scope: Users
lldap  | 2024-02-19T18:56:32.279744623Z 2024-02-19T18:56:32.189226524+00:00  DEBUG    │  │  ┝━ get_user_list [ 935µs | 0.02% / 0.44% ]
lldap  | 2024-02-19T18:56:32.279750738Z 2024-02-19T18:56:32.189257385+00:00  DEBUG    │  │  │  ┝━ 🐛 [debug]:  | filters: And([And([And([]), And([And([]), MemberOf("ocis_users")]), Not(UserId(UserId("false")))]), UserId(UserId("adam"))])
lldap  | 2024-02-19T18:56:32.279757960Z 2024-02-19T18:56:32.189265069+00:00  DEBUG    │  │  │  ┕━ list_users [ 885µs | 0.42% ] filters: Some(And([And([And([]), And([And([]), MemberOf("ocis_users")]), Not(UserId(UserId("false")))]), UserId(UserId("adam"))])) | _get_groups: false
lldap  | 2024-02-19T18:56:32.279768669Z 2024-02-19T18:56:32.190937607+00:00  DEBUG    │  │  │     ┕━ 🐛 [debug]:  | return: [UserAndGroups { user: User { user_id: UserId("adam"), email: "adam@example.nz", display_name: Some("Adam Shand"), creation_date: 2023-05-19T09:37:38.282695250, uuid: Uuid("c5c8724a-bfaa-36d0-ae77-6cc936bf5808"), attributes: [AttributeValue { name: "avatar", value: Serialized("hash: 0xEDB02F50A5FB7F1D") }, AttributeValue { name: "first_name", value: Serialized("Adam") }, AttributeValue { name: "last_name", value: Serialized("Shand") }] }, groups: Some([GroupDetails { group_id: GroupId(5), display_name: "calibre", creation_date: 2023-05-20T05:15:14.956410978, uuid: Uuid("59463dc6-5700-3d0d-96bc-432a5aa43f5f"), attributes: [] }, GroupDetails { group_id: GroupId(4), display_name: "jellyfin-admin", creation_date: 2023-05-20T03:35:06.456309899, uuid: Uuid("e5c0de8a-25a7-37f1-be06-eb82b97e7d6d"), attributes: [] }, GroupDetails { group_id: GroupId(1), display_name: "lldap_admin", creation_date: 2023-05-19T09:35:39.078795764, uuid: Uuid("100fcf69-388b-3725-889f-8c496f9cec40"), attributes: [] }, GroupDetails { group_id: GroupId(6), display_name: "ocis_users", creation_date: 2024-02-19T01:08:57.207719922, uuid: Uuid("4afbf888-08c1-336a-8c79-45a7c4570060"), attributes: [] }]) }]
lldap  | 2024-02-19T18:56:32.279785740Z 2024-02-19T18:56:32.191568379+00:00  DEBUG    │  │  ┝━ expand_attribute_wildcards [ 13.3µs | 0.01% ] ldap_attributes: ["dn", "uid", "mail", "displayName", "sn", "givenName", "entryuuid", "text"]
lldap  | 2024-02-19T18:56:32.279792693Z 2024-02-19T18:56:32.191581211+00:00  DEBUG    │  │  │  ┕━ 🐛 [debug]:  | resolved_attributes: ["dn", "uid", "mail", "displayName", "sn", "givenName", "entryuuid", "text"]
lldap  | 2024-02-19T18:56:32.279798975Z 2024-02-19T18:56:32.191595559+00:00  WARN     │  │  ┕━ 🚧 [warn]: Ignoring unrecognized group attribute: text\n\
lldap  | 2024-02-19T18:56:32.279805325Z                       To disable this warning, add it to "ignored_user_attributes" in the config.
```

File: example_configs/ocis (hunk 2 above)
Comment (adamshand) [reply example_configs/ocis:90]:
```
lldap  | 2024-02-19T19:07:27.839235219Z 2024-02-19T19:07:27.251326853+00:00  INFO     LDAP session [ 146ms | 0.15% / 100.00% ]
lldap  | 2024-02-19T19:07:27.839248904Z 2024-02-19T19:07:27.251492998+00:00  INFO     ┝━ LDAP request [ 146ms | 0.10% / 99.79% ]
lldap  | 2024-02-19T19:07:27.839255914Z 2024-02-19T19:07:27.251520215+00:00  DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 1, op: BindRequest(LdapBindRequest { dn: "uid=admin,ou=people,dc=example,dc=nz", cred: LdapBindCred::Simple }), ctrl: [] }
lldap  | 2024-02-19T19:07:27.839294785Z 2024-02-19T19:07:27.251526304+00:00  DEBUG    │  ┝━ do_bind [ 146ms | 0.04% / 99.69% ] dn: uid=admin,ou=people,dc=example,dc=nz
lldap  | 2024-02-19T19:07:27.839303723Z 2024-02-19T19:07:27.251547372+00:00  DEBUG    │  │  ┝━ bind [ 145ms | 0.03% / 99.44% ]
lldap  | 2024-02-19T19:07:27.839309994Z 2024-02-19T19:07:27.251556463+00:00  DEBUG    │  │  │  ┝━ get_password_file_for_user [ 239µs | 0.16% ] user_id: UserId("admin")
lldap  | 2024-02-19T19:07:27.839316609Z 2024-02-19T19:07:27.252077940+00:00  DEBUG    │  │  │  ┕━ passwords_match [ 145ms | 99.25% ] username: admin
lldap  | 2024-02-19T19:07:27.839323348Z 2024-02-19T19:07:27.397189311+00:00  DEBUG    │  │  ┝━ get_user_groups [ 301µs | 0.21% ] user_id: "admin"
lldap  | 2024-02-19T19:07:27.839330134Z 2024-02-19T19:07:27.397680128+00:00  DEBUG    │  │  │  ┕━ 🐛 [debug]:  | return: {GroupDetails { group_id: GroupId(1), display_name: "lldap_admin", creation_date: 2023-05-19T09:35:39.078795764, uuid: Uuid("100fcf69-388b-3725-889f-8c496f9cec40"), attributes: [] }, GroupDetails { group_id: GroupId(5), display_name: "calibre", creation_date: 2023-05-20T05:15:14.956410978, uuid: Uuid("59463dc6-5700-3d0d-96bc-432a5aa43f5f"), attributes: [] }, GroupDetails { group_id: GroupId(6), display_name: "ocis_users", creation_date: 2024-02-19T01:08:57.207719922, uuid: Uuid("4afbf888-08c1-336a-8c79-45a7c4570060"), attributes: [] }, GroupDetails { group_id: GroupId(4), display_name: "jellyfin-admin", creation_date: 2023-05-20T03:35:06.456309899, uuid: Uuid("e5c0de8a-25a7-37f1-be06-eb82b97e7d6d"), attributes: [] }}
lldap  | 2024-02-19T19:07:27.839377439Z 2024-02-19T19:07:27.397690259+00:00  DEBUG    │  │  ┕━ 🐛 [debug]: Success!
lldap  | 2024-02-19T19:07:27.839385521Z 2024-02-19T19:07:27.397698008+00:00  DEBUG    │  ┕━ 🐛 [debug]:  | response: BindResponse(LdapBindResponse { res: LdapResult { code: Success, matcheddn: "", message: "", referral: [] }, saslcreds: None })
lldap  | 2024-02-19T19:07:27.839392663Z 2024-02-19T19:07:27.397966112+00:00  INFO     ┕━ LDAP request [ 86.5µs | 0.06% ]
lldap  | 2024-02-19T19:07:27.839399019Z 2024-02-19T19:07:27.397990976+00:00  DEBUG       ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 2, op: AddRequest(LdapAddRequest { dn: "uid=ocis_disabled,ou=groups,dc=example,dc=nz", attributes: [LdapPartialAttribute { atype: "objectClass", vals: ["groupOfNames", "top"] }, LdapPartialAttribute { atype: "member", vals: [""] }] }), ctrl: [] }
lldap  | 2024-02-19T19:07:27.839406575Z 2024-02-19T19:07:27.398005411+00:00  DEBUG       ┕━ 🐛 [debug]:  | response: AddResponse(LdapResult { code: InvalidDNSyntax, matcheddn: "", message: "Unexpected DN format. Got \"uid=ocis_disabled,ou=groups,dc=example,dc=nz\", expected: \"uid=id,ou=people,dc=example,dc=nz\"", referral: [] })



```

File: example_configs/ocis (hunk 1 above)
Comment (nitnelave) [reply example_configs/ocis:69]: @FirstYear, I think we got ourselves a parsing bug in ldap3_proto: it seems they're trying to use an attribute option (like in http://www.ldapexplorer.com/en/manual/107070400-binary-attributes.htm) but we're parsing that as an extra attribute. I'm not sure where exactly in the rfc it's described, but the attribute "uuid;text" should resolve to just "uuid" with the text option.

For this PR, no action item, we'll merge the config as is.

File: example_configs/ocis (hunk 2 above)
Comment (nitnelave) [reply example_configs/ocis:90]: I see, they're going for "ask for forgiveness, not permission": they're trying to create the group without checking whether it exists (or apparently whether their config allows them to write to the LDAP server).

We have partial support for creating _users_ via LDAP, but not groups, hence the (bad) error message about an invalid DN (it expects a user DN).

It's not crucial for this integration, but we could add support for creating groups (there's an issue for better LDAP support, #518).

**Review by nitnelave — APPROVED** [review]

**Review by nitnelave — APPROVED** [review]

**Comment by nitnelave** [comment]
A couple of small notes: the file should be ocis.md (you're missing the extension). And can you add it to the list of integrations in the readme?

**Comment by adamshand** [comment]
Fixed.

**Comment by nitnelave** [comment]
Thanks!
